p2p/swarm.go

```diff
@@ -444,76 +495,281 @@ func (s *swarm) onRemoteClientMessage(msg net.IncomingMessageEvent) error {
+	if ok := timesync.CheckMessageDrift(pm.Metadata.Timestamp); !ok {
+		// TODO: consider kill connection with this node and maybe blacklist
+		// TODO : Also consider moving send timestamp into metadata(encrypted).
+		return ErrOutOfSync
 	}
 
-	s.lNode.Debug("Authorized %v protocol message ", pm.Metadata.Protocol)
+	s.lNode.Debug("Authorized %v protocol message ", pm.Metadata.NextProtocol)
 
 	remoteNode := node.New(msg.Conn.RemotePublicKey(), "") // if we got so far, we already have the node in our rt, hence address won't be used
 	// update the routing table - we just heard from this authenticated node
 	s.dht.Update(remoteNode)
 
-	// participate in gossip even if we don't know this protocol
-	if pm.Metadata.Gossip { // todo : use gossip uid
-		s.LocalNode().Debug("Got gossip message! relaying it")
-		// don't block anyway
-		err = s.gossip.Broadcast(decPayload) // err only if this is an old message
-	}
+	// route authenticated message to the registered protocol
 
-	if err != nil {
-		return nil
+	var data service.Data
+
+	if payload := pm.GetPayload(); payload != nil {
+		data = service.Data_Bytes{Payload: payload}
+	} else if wrap := pm.GetMsg(); wrap != nil {
+		data = service.Data_MsgWrapper{Req: wrap.Req, MsgType: wrap.Type, ReqID: wrap.ReqID, Payload: wrap.Payload}
 	}
+
+	return s.ProcessProtocolMessage(remoteNode, pm.Metadata.NextProtocol, data)
+}
+
+// ProcessProtocolMessage passes an already decrypted message to a protocol.
+func (s *swarm) ProcessProtocolMessage(sender node.Node, protocol string, data service.Data) error {
 	// route authenticated message to the reigstered protocol
 	s.protocolHandlerMutex.RLock()
-	msgchan := s.protocolHandlers[pm.Metadata.Protocol]
+	msgchan := s.protocolHandlers[protocol]
 	s.protocolHandlerMutex.RUnlock()
 
 	if msgchan == nil {
-		s.LocalNode().Errorf("there was a bad protocol ", pm.Metadata.Protocol)
 		return ErrNoProtocol
 	}
+	s.lNode.Debug("Forwarding message to %v protocol", protocol)
 
-	s.lNode.Debug("Forwarding message to protocol")
+	msgchan <- protocolMessage{sender, data}
 
-	if payload := pm.GetPayload(); payload != nil {
-		msgchan <- protocolMessage{remoteNode, service.Data_Bytes{Payload: payload}}
-	} else if wrap := pm.GetMsg(); wrap != nil {
-		msgchan <- protocolMessage{remoteNode, service.Data_MsgWrapper{Req: wrap.Req, MsgType: wrap.Type, ReqID: wrap.ReqID, Payload: wrap.Payload}}
+	return nil
+}
+
+// Broadcast creates a gossip message signs it and disseminate it to neighbors.
+func (s *swarm) Broadcast(protocol string, payload []byte) error {
+	return s.ProcessProtocolMessage(s.lNode.Node, s.gossip.ProtocolName, service.Data_Bytes{payload})
+}
+
+// Neighborhood : neighborhood is the peers we keep close , meaning we try to keep connections
+// to them at any given time and if not possible we replace them. protocols use the neighborhood
+// to run their logic with peers.
+
+func (s *swarm) publishNewPeer(peer crypto.PublicKey) {
+	s.newPeerLock.RLock()
```

Comment:
you don't really need two locks (new and del) you can use the same lock for both (since you're only calling read lock)